File: tessoku/python/A29.py
def power(a, b, m):
    ans = 1
    p = a
    for i in range(len(bin(b)) - 2):
        if b & (1 << i):
            ans = (ans * p) % m
        p = (p * p) % m

    return ans


a, b = map(int, input().split())
print(power(a, b, int(1e9) + 7))

File: tessoku/rust/A29.rs
use proconio::input;

fn main() {
    input! {
        a: usize,
        b: usize,
    }

    println!("{}", power(a, b, 1000_000_007));
}

fn power(a: usize, mut b: usize, m: usize) -> usize {
    let mut base = a;
    let mut res = 1;

    while b > 0 {
        if b % 2 == 1 {
            res *= base;
            res %= m;
        }
        base *= base;
        base %= m;
        b /= 2;
    }

    res
}
